039-auto-rebalance-cluster-scaling.md

```diff
@@ -0,0 +1,264 @@
+  cruiseControl:
+    # ...
+    autoRebalance:
+      # using a specific template when adding brokers
+      - mode: add-brokers
+        template: 
+          name: my-add-brokers-rebalancing-template
+      # using a different template when removing brokers  
+      - mode: remove-brokers
+        template: 
+          name: my-remove-brokers-rebalancing-template
+      # NOTE: no full auto-rebalancing   
+```
+
+The user could even decide to have auto-rebalancing only when adding or removing brokers but not for both.
+
+This way provides the greatest flexibility to describe in which case to run the auto-rebalancing and with which configuration.
+
+### Auto-rebalancing on scaling up
+
+Following the operator flow when auto-rebalancing on scaling up.
+
+1. The user scales up the cluster by increasing the `spec.kafka.replicas` field.
+2. The `KafkaReconciler` takes care of it by scaling up the cluster as usual, by adding additional brokers.
+3. If the `spec.cruiseControl.autoRebalance` contains an entry with `add-brokers` mode, the `KafkaReconciler` creates a `KafkaRebalance` custom resource named as `<cluster-name>-auto-rebalance-add-brokers` by using the corresponding template and sets:
+   * the `spec.mode` as `add-brokers` and the `spec.brokers` with the list of brokers already added to the cluster.
+   * the `strimzi.io/rebalance-auto-approval=true` annotation as described by proposal [038](https://github.com/strimzi/proposals/blob/main/038-optimization-proposal-autoapproval.md).
+4. The `KafkaRebalanceAssemblyOperator` detects the new rebalancing resource and acts as usual by requesting Cruise Control to generate an optimization proposal via the REST API on the `/add_broker` endpoint.
+5. When the optimization proposal is ready, the `KafkaRebalanceAssemblyOperator` approves it automatically, because of the `strimzi.io/rebalance-auto-approval=true` annotation, by annotating the `KafkaRebalance` custom resource with the `strimzi.io/rebalance=approve` annotation as defined by proposal [038](https://github.com/strimzi/proposals/blob/main/038-optimization-proposal-autoapproval.md).
+6. If/When the rebalancing ends successfully, the corresponding `KafkaRebalance` custom resource is deleted by the `KafkaReconciler`.
+
+#### Error handling
+
+During the above flow, errors can happen after the scaling up:
+
+* There is an error when getting the proposal.
+* After auto-approving the proposal, there is an error when running the rebalance.
+
+In both scenarios, the reconciliation ends with the `KafkaRebalance` custom resource in the `NotReady` state as per the `KafkaRebalanceAssemblyOperator` states machine.
+The cluster operator logs a warning about the rebalancing not happened.
+It's the user who should deal with this scenario:
+
+* Checking the reason why the `KafkaRebalance` is in the `NotReady` state and deleting the custom resource.
+* If the user still wants rebalancing, they have to create a new `KafkaRebalance` custom resource manually.
```

Comment:
Presumably the user could try to reuse the `KafkaRebalance` by adding a refresh annotation? What would happen in that case?

Reply:
It should just work. There is nothing preventing the user to put his hands on a `KafkaRebalance` created by the operator unless we could set up some RBAC for it?